### PR TITLE
[FLINK-32618][jdbc-driver] Remove flink-core dependency from jdbc module

### PR DESCRIPTION
**Pull request by fsk119**
## What is the purpose of the change

*Currently the jdbc bundle jar also includes flink-core and sql-client jar. However, the jdbc bundle relies on flink-core is because we use AutoCloseableRegistry in the ExecutorImpl. So we copy the class from the flink-core and remove the dependency.*

*I think the jdbc-bundle jar is used in the java program, which provides a client to communicate with gateway. Sql-client prodives a terminal client for users to submit their sql easily. So I think it should the sql-client relies on the jdbc, which is different from the current implementation.*

## Brief change log

  - *Copy the AutoCloseableRegistry from the flink-core to jdbc module*
  - *Move the Executor from the sql-client module to jdbc module. The Executor provides a low level API to submit sql to the gateway comparing to the jdbc.*
  - *Change the dependency between jdbc and sql-client.*

### CODE REVIEW

**Comment by fsk119** [comment]
@flinkbot run azure

**Review by zentol — COMMENTED** [review]
Why did you bundle flink-core in the first place? Why not just use the flink-core that flink provides?

File: flink-table/flink-sql-jdbc-driver/src/main/java/org/apache/flink/table/gateway/AutoCloseableRegistry.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.gateway;
+
+import org.apache.flink.util.IOUtils;
```
Comment (zentol) [inline flink-table/flink-sql-jdbc-driver/src/main/java/org/apache/flink/table/gateway/AutoCloseableRegistry.java:21]: this is a flink-core dependency

**Comment by fsk119** [comment]
Hi @zentol. The reason we bundle flink-core is to make the tool easier to use. They can just add the jdbc-bundle dependency in their projects without detailed dependency management.

I think it is difficult for jdbc users to use flink-dist. As the example "FlinkDriverExample" shows, users only need to write java code to communicate with the running remote gateway service without building flink-dist in the local environment.

Your suggestion about the usage of the IOUtils makes me realize it's difficult to remove the flink-core... Many dependencies used by the flink-jdbc-bundle use `org.apache.flink.util.Preconditions` everywhere, e.g. `RestClient`. I think it's not possible to remove this directly.

**Comment by fsk119** [comment]
The PR only decrease the bundle jar from 21MB to 18MB because we can not remove the flink-core easily. On the other hand, many users may use the ExecutorImpl to build their own platform. The change in this PR modifies the class package. So it's better to delay the change until we can remove the flink-core.

**Pull request closed without merging**